Cached the calculation of the field's label and source attributes

thedrow opened the pull request:
This results in much faster binding times and less initialization overhead even between different serializers who use the same source.

tomchristie commented on rest_framework/compat.py:
```diff
@@ -168,6 +168,174 @@ def __init__(self, *args, **kwargs):
+            return update_wrapper(wrapper, user_function)
+
+        return decorating_function
+
```
Wouldn't be happy with us pulling in this much compat stuff for an unverified improvement.
Happy to consider alternatives.

tomchristie commented:
Closing as described.

thedrow commented:
Since we dropped 1.4 support and all Django versions now include the lru_cache compatibility layer we should reconsider this PR.
@tomchristie what do you think? I can edit it to remove the branch that vendors it if it doesn't exist in Django.

xordoquy commented:
@thedrow Do you have some numbers about how much it affects performances ?
We're not going to accept changes that make code harder to maintain just for the sake of it.

thedrow commented:
Ok so let's review the changes I introduced here:
```python
return field_name.replace('_', ' ').capitalize()
```
performs two loops, one for replacing the string's characters and one to capitalize the string. Each comes with it's own memory allocations as opposed to just one memory allocation and a lookup which is obviously faster, even without running any benchmarks.
```python
if source == '*':
        return ()
    return tuple(source.split('.'))
```
Has a code branch that is completely skipped once it has been used once since it's cached and requires only a lookup instead of memory access, comparison to a string. At least in CPython empty tuples are a singleton so no memory allocation is required.
The other code branch requires us to run two loops and, one that splits the string and allocates a list of strings (so N+1 allocations where N is how many '.' characters are present in the string) and conversion to a tuple which requires us to run the second loop and allocate memory for the tuple.
This code is obviously faster when a serializer is used more than once which is almost always (or even always) the case.
Does it really matter by how much?
If it really does here are the microbenchmarks for ```get_field_label()```:
Without caching:
```
%timeit get_field_label('first_name')
The slowest run took 9.52 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 300 ns per loop
```
With caching:
```
%timeit get_field_label('first_name')
The slowest run took 10.47 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 2.28 µs per loop
```
Even ~20µs which was the slowest run in the cached version is much much faster than 300 ns.
And here are the microbenchmarks for ```get_source_attributes()```:
Without caching:
```
%timeit get_source_attributes('user.first_name')
The slowest run took 7.01 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 442 ns per loop
```

With caching:
```
%timeit get_source_attributes('user.first_name')   
The slowest run took 16.55 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 2.31 µs per loop
```

So in total you save almost a milisecond which does make a difference if your goal is to build a responsive API that responds in 100ms or less (or even 50ms or less like github does).
Not to mention that PyPy will JIT the entire function call to something else since it will be able to tell that the value is already cached.

@xordoquy Does that satisfy you?

xordoquy commented:
Thanks for the explanations.
Before reading them carefully, I'd just want to put some emphasis on a point.

> So in total you save almost a milisecond which does make a difference if your goal is to build a responsive API that responds in 100ms or less (or even 50ms or less like github does).

Note that DRF goal isn't to perform as well as a finely tuned API. Though better performance is always a good thing DRF politic is to keep the code base as maintainable as possible. We mostly do that on our free time and the fastest it is the better. I'm not assuming anything on your patch or the improvement - just wanted to make this point clear.

thedrow commented:
@xordoquy I am aware of that, and I disagree with that mission statement as I'm sure that @tomchristie is aware. APIs created with DRF are not toy-projects. Actual startups and companies use your code.
Maintainable public API is a better goal in my opinion. Not introducing a new utility function that runs faster just because it complicates the code a little bit isn't a tradeoff I'd accept. This is one of the reasons why I don't volunteer for to maintain the project.

xordoquy commented:
So, I ran a couple of serializer performance benchmark with and without the patch. I hardly noticed any change meaning some results are better other not so.

Without the patch:
```
----------- benchmark: 28 tests, min 5 rounds (of min 25.00us), 1.00s max time, timer: time.perf_counter -----------
Name (time in us)                                    Min           Max          Mean      StdDev  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------
test_model_serializer[1]                       1525.2780    21322.6000     1797.4791    971.2865     443           1
test_model_serializer[10]                      2051.7290     5157.8860     2430.2919    401.2175     448           1
test_model_serializer[100]                     7441.5250    36466.4750    10135.1612   5023.1656      56           1
test_model_serializer[1000]                   63451.7840   146065.8100    84354.0339  26261.3873      14           1
test_model_serializer[10000]                 811774.2290   830105.0250   818217.7870   6971.1577       5           1
test_serializer[1]                              543.8980     1644.0060      673.6386    152.3511    1306           1
test_serializer[10]                             983.4580     3155.0970     1227.6596    280.6953     707           1
test_serializer[100]                           5297.7590   123935.1800     7419.1381   9317.6831     181           1
test_serializer[1000]                         52058.8130   103501.1050    66437.2662  19607.4632      19           1
test_serializer[10000]                       661798.9990   714868.6920   695493.2190  19925.4259       5           1
test_model_deserializer[1]                     1786.8360    88047.8740     2280.3400   4300.3315     402           1
test_model_deserializer[10]                    2689.1860     4853.6730     3134.4870    476.3735     325           1
test_model_deserializer[100]                  11397.0250    41194.4530    13731.8541   4441.7180      69           1
test_model_deserializer[1000]                102763.6310   155265.3670   120661.2215  19501.1926      10           1
test_model_deserializer[10000]              1210716.7170  1262424.8150  1240261.6800  26084.2464       5           1
test_deserializer[1]                            651.9440    96472.6810      873.8347   2727.9485    1236           1
test_deserializer[10]                          1520.5480     3598.6620     1815.6627    338.4849     619           1
test_deserializer[100]                        10088.2250    40872.9290    12381.4937   5218.6962      85           1
test_deserializer[1000]                      102730.5230   148157.7430   114231.2622  16429.2835      10           1
test_deserializer[10000]                    1161094.9840  1212289.9930  1183939.7870  24258.4315       5           1
test_model_serializer_field_qt[1]               322.2780     1227.3670      368.5928     78.7642    1477           1
test_model_serializer_field_qt[10]              887.1220     2740.5120     1108.0492    233.9420     895           1
test_model_serializer_field_qt[100]            6397.2660   109769.2190     8579.0298   8927.7186     143           1
test_model_serializer_field_qt[1000]          66112.5000   116169.0830    79603.1424  17108.4881      14           1
test_model_serializer_field_qt_10[1]            425.0460    51234.3330      532.6065   1251.8867    1657           1
test_model_serializer_field_qt_10[10]          1302.1750     4337.4880     1597.4888    369.5414     596           1
test_model_serializer_field_qt_10[100]        10165.4000    43799.2050    12955.2116   6189.3264      86           1
test_model_serializer_field_qt_10[1000]      113506.4890   160436.7580   132493.9520  21672.1105       9           1
--------------------------------------------------------------------------------------------------------------------
```

With the patch:
```
----------- benchmark: 28 tests, min 5 rounds (of min 25.00us), 1.00s max time, timer: time.perf_counter -----------
Name (time in us)                                    Min           Max          Mean      StdDev  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------
test_model_serializer[1]                       1537.9040     3364.0860     1765.3033    258.4261     433           1
test_model_serializer[10]                      2056.1190    20733.1930     2477.5460    983.9026     418           1
test_model_serializer[100]                     7250.4320    35344.9640     9278.6771   4671.9529     111           1
test_model_serializer[1000]                   61215.7420   113284.1990    78849.5944  20955.8997      16           1
test_model_serializer[10000]                 803446.8880   829424.6170   816238.6926  10493.9992       5           1
test_serializer[1]                              568.5560     1589.2170      703.4937    135.3298    1254           1
test_serializer[10]                            1020.6760     2861.2380     1273.9852    244.3607     675           1
test_serializer[100]                           5443.1100   114162.3210     7326.7888   9408.4830     145           1
test_serializer[1000]                         53052.5910    97966.9310    64394.9718  17611.8905      17           1
test_serializer[10000]                       675793.1380   735761.0080   707258.6252  23228.7305       5           1
test_model_deserializer[1]                     1812.9610     4574.2720     2040.2489    340.3409     412           1
test_model_deserializer[10]                    2706.4190     5801.4520     3085.9699    499.9377     317           1
test_model_deserializer[100]                  11111.4480    16608.9700    12753.6041   1227.5504      78           1
test_model_deserializer[1000]                 98034.0410   144884.6090   110411.5890  19491.2105       5           1
test_model_deserializer[10000]              1112129.1550  1181008.4020  1145818.7670  31803.3287       5           1
test_deserializer[1]                            671.8600    95935.9000      875.1965   2842.2005    1124           1
test_deserializer[10]                          1522.0350     4249.8260     1809.9892    337.4832     525           1
test_deserializer[100]                         9886.0450    37700.6020    11948.1214   4570.4764      97           1
test_deserializer[1000]                       96022.6950   140588.9770   111099.2636  18600.2465      10           1
test_deserializer[10000]                    1138322.7460  1206744.5350  1172794.8708  25554.5611       5           1
test_model_serializer_field_qt[1]               330.3570     1237.1860      374.3638     81.7392    1666           1
test_model_serializer_field_qt[10]              907.6910    97441.9010     1254.7376   3609.5032     715           1
test_model_serializer_field_qt[100]            6514.1930    34704.8990     8201.6360   3441.2067     117           1
test_model_serializer_field_qt[1000]          71186.7040   121653.6950    82548.0291  17818.1180      14           1
test_model_serializer_field_qt_10[1]            430.1880    42167.6090      526.0886   1054.5525    1578           1
test_model_serializer_field_qt_10[10]          1323.8260     3054.5430     1606.1547    306.0455     668           1
test_model_serializer_field_qt_10[100]        10226.6390    43658.2840    12985.8670   5809.3525      31           1
test_model_serializer_field_qt_10[1000]      109994.9660   154181.9530   129855.6177  18968.2633       9           1
--------------------------------------------------------------------------------------------------------------------
```

thedrow commented:
I think that's because you're not benchmarking the field binding which is done when you call https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/serializers.py#L330.
It could be that by the time you're running the benchmark the fields are already bound.

xordoquy commented:
The time includes the serializer instantiation which means that the fields are not bound before the timer starts.

thedrow commented:
@xordoquy Where can I see your benchmark code?

xordoquy commented:
It's available there: http://github.com/xordoquy/django-rest-framework-benchmark

thedrow commented:
I'll try it myself and I'll let you know.
